pyresample/gradient/__init__.py

```diff
@@ -618,7 +353,7 @@ def block_bilinear_interpolator(data, indices_xy, fill_value=np.nan, block_info=
     res = ((1 - weight_l) * (1 - weight_p) * data[..., l_start, p_start] +
            (1 - weight_l) * weight_p * data[..., l_start, p_end] +
            weight_l * (1 - weight_p) * data[..., l_end, p_start] +
-           weight_l * weight_p * data[..., l_end, p_end])
+           weight_l * weight_p * data[..., l_end, p_end]).astype(data.dtype)
```

Comment:
Should we make sure the calculation is does not promote the `dtype` in the first place by making the weights the same type?

Reply:
I'm not sure. I think I want to keep the highest precision to do the bilinear interpolation, and only then convert to float32. But maybe I'm being too careful?

Reply:
I doubt there's nothing lost whether the interpolation is done with `float32` or `float64`. Of course the weights need to be floats, so casting to integer dtypes would still need additional handling... But as a test, it might be interesting to see the difference in memory usage between f32/f64 weights.

Reply:
In the case of bilinear interpolation, the floating index is indeed only between 0 and 1, but in general for the gradient search we use that type also for the gradients and the coordinates in general. And I think that for that float64 makes more sense, right? Then we just take the fractional part and use that for the interpolation.
Also, that index-building function has no knowledge of the data type. What I can do is make the weights float32s before that line if you want. I think it's all we can do right now...

Reply:
I was thinking that switching `weight_l` and `weight_p` for this bit to `data.dtype` or `float64`(?) for integer data wouldn't affect anything else? This is the part where the result is calculated, so only should affect anything else. Or am I completely lost?

Reply:
You got it right, I thought you wanted to change the cython code to handle all indices in float32 when data was float32 for example, but we actually came to the same conclusion. I'll fix this soon.

Reply:
Fixed in [`4dd2948` (#626)](https://github.com/pytroll/pyresample/pull/626/commits/4dd294877c9c24a69890d7634cefeb1f7c9c0270)